### ISSUE TITLE
What operations do we actually need? 

### ISSUE THREAD

**Issue by dfdx** [denizyuret/CUDNN.jl#8]
I've just "implemented" (by copy-and-pasting from Knet.jl) a new [CuConv.jl](https://github.com/dfdx/CuConv.jl) package for convolution and pooling on CUDA. I haven't tested it thoroughly yet, but all in all it seems to work as is and it basically covers everything that I personally want to have from cuDNN. Also, as discussed in JuliaGPU/CuArrays.jl#22, many operations from cuDNN might be better to implement in Julia itself. So the question arises - what other functions do we actually need? 

What's included in Knet.jl / CuConv.jl: 

1. Convolution.
2. Deconvolution.
3. Pooling.
4. Unpooling.

What's not there:

1. LRNCrossChannel (I don't actually know what it is).
2. Divisive normalization.
3. Batch normalization (not implemented in CUDNN.jl, but there's [version](https://github.com/denizyuret/Knet.jl/blob/c034108dd4f7fcb3c8431224edf904a150d5d334/examples/resnet.jl#L134-L152) from Knet).
4. TransformTensor (`alpha * src + beta * dest -> dest`).
5. AddTensor.
6. SetTensor (`src .= value`).
7. ScaleTensor (`src .*= alpha`).
8. Activation functions (sigmoid, relu, tanh).
9. Softmax.

For me almost everything here looks like a better fit for CUBLAS + CUDAnative (with additional bonus of fused broadcasting). If so, we could deprecate CUDNN and instead concentrate on a more Julian approach. 

cc: @MikeInnes @jekbradbury @denizyuret

**Comment by jekbradbury** [comment]
Cool! Yeah, I think everything else would be best left to Julia, with the possible (but probably not worth it) exception of batchnorm (just because the implementation in cuDNN 7 is exceptionally efficient while batchnorm can sometimes be a bottleneck for convnets) and later (probably only after the private persistent kernel codegen API is made public) the persistent-kernel RNNs.
LRN is Local Response Normalization, an outdated predecessor of batchnorm/layernorm that was used in the original AlexNet paper but is a) complicated and b) not used any more except for AlexNet replications.

**Comment by denizyuret** [comment]
In general I think we should run benchmarks before retiring cudnn versions.
4-8 are regular broadcast ops. 9 softmax is generally very expensive, if
cudnn has an efficient version we should use it. There is also lstm,
batchnorm etc. that were added to cudnn after I last updated cudnn.jl which
may be worth benchmarking.

**Comment by jekbradbury** [comment]
For softmax I believe [this kernel](https://github.com/pytorch/pytorch/pull/2899) is competitive with cuDNN, and we should be able to port something similar to CUDAnative

**Comment by dfdx** [comment]
> There is also lstm, batchnorm etc.

Sounds interesting, I'll take a look at API. 

Then the plan for me now is:

1. Make CuConv.jl usable (tests + other minor issues) so to enable other people to use convolutions in their code and get early feedback. 
2. Check out batchnorm, LSTM, softmax, etc. 
3. If I get something close to a full cuDNN, replace the current implementation with the new one and drop CuConv.jl. 

I assume that CUDNN.jl in its current state isn't actively used and dependent packages may be easily updated.

Later we can test alternative kernels (like the one for softmax) and e.g. put both with different names (`softmax` vs. `softmax_native`, for example).

**Comment by denizyuret** [comment]
I think that assumption is correct -- I haven't updated the package for the
last two releases of the cudnn library, so I doubt a lot of people are
using it.  Even if they are that's what versioning is for :).  Getting it
compatible and complete wrt the last version released by NVIDIA would be a
good goal.  A lot of the code can be automatically generated from the
header files.  The only exception is the high level interface, which needs
to be hand fixed.

**Comment by MikeInnes** [comment]
Sounds like a great plan!

For CuConv.jl you might want to use CUDAapi and CUDAdrv in place of manually finding cuda (you may not need direct access to the runtime at all). Those are pretty robust at this point. You don't even need to find CUDNN yourself, just grab `CuArrays.libcudnn`.

**Comment by Evizero** [comment]
There are also 

- `cudnnRNNForwardInference`
- `cudnnRNNForwardTraining`
- `cudnnRNNBackwardData`
- `cudnnRNNBackwardWeights`

see 

- http://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#tensor_ops)
- http://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnRNNMode_t

**Comment by dfdx** [comment]
`conv2d` and `pool`, as well as their gradients, are now checked and ready for use in CuConv.jl. To test their correctness, I also extracted CPU versions from Knet.jl to NNlib.jl (see FluxML/NNlib.jl#2). I still need to implement / test `deconv` and `unpool`, but first I want to check out batchnorm, RNNs, etc. to see how long it may take.

**Comment by dfdx** [comment]
`batchnorm` is [almost here](https://github.com/dfdx/CuConv.jl/blob/batchnorm/src/batchnorm.jl), but I have doubts about the interface. Unlike e.g. convolution, batch normalization is a stateful operation. In particular, in addition to input tensor `x` it: 

 * learns internal scale and bias (gamma and beta) parameters
 * passes to backward operation running mean and variance as well as cache for intermediate results

TensorFlow stores these parameters in the current graph. PyTorch saves them as module parameters and then uses in forward and backward passes. Since CUDNN.jl is a general-purpose library, we neither graphs, nor modules. 

I see the following options:

 * pass states explicitly, e.g.:

```
states = batchnorm_states()
y = batchnorm(x, states)   # modifies states
dy = ...
dx = batchnorm_grad(x, dy, states)
```

 * return tuple of `(y, internal_states)`,  e.g.:

```
params = batchnorm_params()  # gamma and beta
y, s = batchnorm(x, params)
dy = ...
dx = batchnorm_grad(x, dy, s)
```
 
 * create a stateful model in CUDNN.jl, e.g.:

```
bn = BatchNorm()
y = forward(bn, x)
dy = ...
dx = backward(bn, x, dy)
```
In my XGrad.jl I can deal with all three, but Flux.jl, Knet.jl and others may require one specific option. Another thing we need to take into consideration is that the same API will have to be implemented for CPU.

**Comment by dfdx** [comment]
And one more thing: batchnorm and dropout behave differently during forward pass in training and inference mode. I'll add a switch, but be prepared to add this capability to your libraries.

**Comment by Evizero** [comment]
In the context of CUDNN.jl my personal opinion is that the interface should be reasonable close to the one of the actual library. If the lowest level interface is already a high-level interface then this may restrict the design decisions of dependent packages.

**Comment by denizyuret** [comment]
The current design has both: an automatically generated low level interface
and a manually designed high level interface.

**Comment by dfdx** [comment]
>  the interface should be reasonable close to the one of the actual library. If the lowest level interface is already a high-level interface then this may restrict the design decisions of dependent packages.

The current interface gives access to nearly all low-level options of wrapped functions, but provides convenient API (e.g. you can pass a single `CuArray` instead of a tensor descriptor and pointer to data, or create and initialize descriptor in 1 call instead of 2 separate functions as in C API). At the same time, sometimes low-level API is too C-centric and it's unclear how a high-level interface should look like (as in case with batch normalization).

**Comment by dfdx** [comment]
I've added batchnorm and softmax (4D), and also renamed the package from CuConv to [CUDNN2](https://github.com/dfdx/CUDNN2.jl). Once FluxML/NNlib#2 is merged, I will be ready to transfer the repo to JuliaGPU organization and register it in METADATA. 

Some details to know:

 * `CUDNN2.softmax4d` is different from `NNlib.softmax` - the former just wraps cuDNN's function and works with 4D tensors while NNlib's version works with 1D and 2D data;
 * NNlib also doesn't provide `batchnorm`, we will need to fix it later to make a unified set of CPU & GPU functions
 * I also tried to implement `dropout`, but all my attempts failed without clear error description; a few failed experiments  may be found in [dropout.jl](https://github.com/dfdx/CUDNN2.jl/blob/master/src/dropout.jl)
 * I also plan to implement RNN stuff at some point, but it requires even larger time investment, so I have to postpone it

**Comment by denizyuret** [comment]
Please see Knet for a recent rnn inplementation and conv optimizations. In
conv apparently using the default algo is not a good idea.

**Comment by dfdx** [comment]
Looks great! I created 2 issues for these changes in CUDNN2.

**Comment by dfdx** [comment]
I've just moved [CUDNN2.jl](https://github.com/JuliaGPU/CUDNN2.jl) to JuliaGPU organization. If you have anything to say about it (e.g. comment the name of purpose), please do it. Otherwise, I'm going to register it in METADATA in a couple of days.

**Comment by MikeInnes** [comment]
Is there any reason not to delete the current cudnn.jl and reuse the name?

**Comment by dfdx** [comment]
There are 2 ways to replace one implementation with another: 

1. Remove one package and rename another to have the same name. But given that CUDNN is registered in METADATA with specific commits, it may significantly break `Pkg.add`/`Pkg.update`. 
2. Wipe out one package and put completely new content instead. This should work, but will completely abuse git workflow and history of commits.

Also CUDNN2.jl isn't complete replacement for CUDNN.jl, there are still a few unique features in the latter. Not sure anybody uses them, but silently removing them does't sound correct to me.

**Comment by MikeInnes** [comment]
You can do a merge commit and keep both histories.

**Comment by dfdx** [comment]
Are you talking about replacing the content? Sure, the history will be kept, but it will be useless. 

Anyway, I don't have strong preference to either option, so let's make a poll: I will post several options, please react with +1 for the one you like most.

**Comment by dfdx** [comment]
1. Register a new package CUDNN2.jl.

**Comment by dfdx** [comment]
2. Replace content of CUDNN.jl.

**Comment by dfdx** [comment]
3. Rename packages (may partially break METADATA).

**Comment by denizyuret** [comment]
As the author of CUDNN I prefer having a single package.  Less confusing
for the user.  We should be able to keep both histories so CUDNN2 becomes
the next release of CUDNN.

**Comment by dfdx** [comment]
[So let it be!](https://github.com/JuliaGPU/CUDNN.jl/pull/9)
